src/coord/src/coord.rs

```diff
@@ -3290,29 +3603,32 @@ impl<S: Append + 'static> Coordinator<S> {
                 instance.replicas_by_id[&replica_id].clone(),
             ));
         }
-
-        self.catalog_transact(Some(session), ops, |_| Ok(()))
-            .await?;
+        self.client_catalog_transact(
+            session,
+            ops,
+            |_| Ok(()),
+            tx,
+            Ok(ExecuteResponse::DroppedComputeInstanceReplicas),
+            Err,
+        )
+        .await;
 
         for (compute_id, replica_id, replica) in replicas_to_drop {
```

Comment:
So how does this work now? The catalog txn happens, these side effects occur, but control to the user session isn't returned until the group commit finishes, which is some unknown amount of time in the future? So this particular session will observe this change as being fully completed (where fully means the builtin table updates are also applied). What happens to other sessions after the time that these side effects are applied and before the builtin table updates are applied? Those sessions would have the catalog report one state (some thing exists or doesn't), but that thing isn't reflected in the builtin tables. Is that a correct description? If so, it feels like a thing we need to either fix or justify.

Reply:
All DDL now takes the following steps, in order:
1. Perform catalog transaction.
2. Drop sources, tables, sinks, indexes, secrets, and replication slots.
3. Submit writes to built in tables to group commit.
4. Create/destroy other resources in STORAGE/COMPUTE (for example dropping the replicas in this loop).
5. Wait for group commit. During this time any other user operation can start and possibly finish depending on the operation.
6. Group commit completes.
7. Return a response to the client.

I think this is OK, because no one will see a partially applied state. Any operation that wants to read the builtin tables, and could potentially see conflicting information compared to the in-memory catalog, would be blocked until their assigned timestamp is closed by group commit finishing. Once the timestamp is closed and they are able to read the builtin table, it will be consistent with the in-memory catalog.